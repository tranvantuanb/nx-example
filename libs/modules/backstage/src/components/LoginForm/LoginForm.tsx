import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useRouter } from 'next/router';
import tw, { styled } from 'twin.macro';
import { Form } from 'antd';
import { useGTMEvent } from '@carro/core/hooks/useGTMEvent';

import {
  CardForm,
  Input,
  Button,
  Alert,
  Checkbox,
  withCardFormContext,
} from '@carro/common/components';

import { setAccessToken } from '@carro/utils/helpers/auth';
import { AuthApi } from '@carro/auth';

const StyledHeader = styled.div`
  ${tw`text-center`}
  .create-one {
    ${tw`text-info`}
  }
`;

interface ComponentProps {
  onLoginSuccess?: (data) => void;
}

const LoginFormInternal: React.FC<ComponentProps> = ({ onLoginSuccess }) => {
  const [form] = Form.useForm();
  const [isDisableSubmitBtn, setIsDisableSubmitBtn] = useState(true);

  const router = useRouter();
  const { t } = useTranslation('auth');
  const [login, { isLoading, isError }] = AuthApi.useLoginByEmailMutation();

  const { pushGTMEvent } = useGTMEvent();

  const handleSubmitForm = async (values) => {
    const res = await login(values).unwrap();
    if (res?.data?.access_token) {
      onLoginSuccess?.(res?.data);

      setAccessToken(res.data.access_token);
      const salesperson = res?.data?.user;

      pushGTMEvent({
        event: 'logged_in',
        page_name: 'login',
        salesperson_id: salesperson.id,
        salesperson_name: salesperson.name,
      });

      const page = (router.query['redirectUrl'] as string) || '/';

      router.push(page);
    }
  };

  const handleFieldValueChange = () => {
    const isExistFieldValue = Object.values(form.getFieldsValue(true)).every(
      (value) => value
    );
    setIsDisableSubmitBtn(!isExistFieldValue);
  };

  const handleRememberMe = (e) => {
    console.log('remember me: ', e.target.checked);
  };

  return (
    <div className="mx-auto w-full max-w-sm lg:w-96">
      <StyledHeader>
        <div className="flex justify-center">
          <img
            className="h-12 w-auto"
            src="/img/veronica-logo.svg"
            alt="veronica"
          />
        </div>
        <div className="mt-8 mb-4 text-[40px] font-bold">Welcome back!</div>
        <div className="mt-8 mb-4 text-[40px] font-bold">
          Welcome back TuanTV!
        </div>
        <div className="text-xs">
          <span>Donâ€™t have an account yet?</span>
          <a href="#" className="ml-2 text-xs font-black text-info">
            Create one!
          </a>
        </div>
      </StyledHeader>
      <div className="mt-8">
        <CardForm
          form={form}
          name="login-form"
          isEditing
          onFinish={handleSubmitForm}
        >
          <CardForm.Row>
            <CardForm.Item
              name="email"
              label={t('Email')}
              colProps={{ span: 24 }}
            >
              <Input
                placeholder={t('Email')}
                onChange={handleFieldValueChange}
              />
            </CardForm.Item>
            <CardForm.Item
              className="password-input"
              name="password"
              label={
                <div className="password">
                  <span>{t('Password')}</span>
                </div>
              }
              colProps={{ span: 24 }}
            >
              <Input
                type="password"
                placeholder={t('Password')}
                onChange={handleFieldValueChange}
              />
            </CardForm.Item>

            <CardForm.Item colProps={{ span: 24 }}>
              <Checkbox onChange={handleRememberMe}>
                {t('Remember me on this device')}!
              </Checkbox>
              <a href="#" className="font-black text-xs text-info float-right">
                {t('Forgot password')}
              </a>
            </CardForm.Item>

            <CardForm.Item colProps={{ span: 24 }} shouldUpdate>
              {() => (
                <Button
                  className="backstage-primary-btn"
                  type="primary"
                  htmlType="submit"
                  loading={isLoading}
                  block
                  disabled={!form.isFieldsTouched(true) || isDisableSubmitBtn}
                >
                  {t('Sign in')}
                </Button>
              )}
            </CardForm.Item>

            <CardForm.Item colProps={{ span: 24 }}>
              {isError && (
                <Alert
                  style={{ marginBottom: '16px' }}
                  message={t('Login failed! Wrong email or password.')}
                  type="error"
                  showIcon
                />
              )}
            </CardForm.Item>
          </CardForm.Row>
        </CardForm>
      </div>
    </div>
  );
};

export const LoginForm = withCardFormContext<ComponentProps>(LoginFormInternal);
