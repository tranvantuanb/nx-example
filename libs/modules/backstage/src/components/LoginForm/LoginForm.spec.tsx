import { render } from '../../test';

import { LoginForm } from './LoginForm';

describe('LoginForm', () => {
  it('should render successfully', () => {
    const { container } = render(<LoginForm />);
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
