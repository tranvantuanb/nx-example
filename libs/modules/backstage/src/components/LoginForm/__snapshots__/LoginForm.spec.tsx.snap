// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render successfully 1`] = `
<div>
  <div
    css="[object Object]"
  >
    <div
      class="css-1a6tac2"
    >
      <div
        css="[object Object]"
      >
        <img
          alt="veronica"
          css="[object Object]"
          src="/img/veronica-logo.svg"
        />
      </div>
      <div
        css="[object Object]"
      >
        Welcome back!
      </div>
      <div
        css="[object Object]"
      >
        <span>
          Donâ€™t have an account yet?
        </span>
        <a
          css="[object Object]"
          href="#"
        >
          Create one!
        </a>
      </div>
    </div>
    <div
      css="[object Object]"
    >
      <form
        class="ant-form ant-form-vertical"
        id="login-form"
      >
        <div
          class="ant-row"
          style="margin: -8px -8px -8px -8px;"
        >
          <div
            class="ant-col ant-col-24"
            style="padding: 8px 8px 8px 8px;"
          >
            <div
              class="ant-row ant-form-item css-9i9zmh-StyledFormItem e1pqsrmx0"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class=""
                  for="login-form_email"
                  title="Email"
                >
                  Email
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <input
                      class="ant-input css-13ito7b-StyledInput e6ii1nn1"
                      id="login-form_email"
                      placeholder="Email"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-24 password-input"
            style="padding: 8px 8px 8px 8px;"
          >
            <div
              class="ant-row ant-form-item css-9i9zmh-StyledFormItem e1pqsrmx0"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class=""
                  for="login-form_password"
                  title=""
                >
                  <div
                    class="password"
                  >
                    <span>
                      Password
                    </span>
                  </div>
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <input
                      class="ant-input css-13ito7b-StyledInput e6ii1nn1"
                      id="login-form_password"
                      placeholder="Password"
                      type="password"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-24"
            style="padding: 8px 8px 8px 8px;"
          >
            <div
              class="ant-row ant-form-item css-9i9zmh-StyledFormItem e1pqsrmx0"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <label
                      class="ant-checkbox-wrapper ant-checkbox-wrapper-in-form-item css-1iv9iew-StyledCheckbox eu8bm9l0"
                    >
                      <span
                        class="ant-checkbox"
                      >
                        <input
                          class="ant-checkbox-input"
                          type="checkbox"
                          value=""
                        />
                        <span
                          class="ant-checkbox-inner"
                        />
                      </span>
                      <span>
                        Remember me on this device
                        !
                      </span>
                    </label>
                    <a
                      css="[object Object]"
                      href="#"
                    >
                      Forgot password
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-24"
            style="padding: 8px 8px 8px 8px;"
          >
            <div
              class="ant-row ant-form-item css-9i9zmh-StyledFormItem e1pqsrmx0"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <button
                      class="ant-btn ant-btn-primary ant-btn-block backstage-primary-btn css-4j3tcx-StyleButton ewiole60"
                      disabled=""
                      type="submit"
                    >
                      <span>
                        Sign in
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-24"
            style="padding: 8px 8px 8px 8px;"
          >
            <div
              class="ant-row ant-form-item css-9i9zmh-StyledFormItem e1pqsrmx0"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
